concordium-contracts-common/src/schema.rs

```diff
@@ -229,6 +277,214 @@ impl FunctionV2 {
     }
 }
 
+/// Describes the schema of an init or a receive function for V1 contracts with
+/// V3 schemas. Differs from [`FunctionV2`] in that a schema for the events can
+/// be included.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum FunctionV3 {
+    Param(Type),
+    /// Rv is short for Return value.
+    Rv(Type),
+    ParamRv {
+        parameter:    Type,
+        return_value: Type,
+    },
+    Error(Type),
+    ParamError {
+        parameter: Type,
+        error:     Type,
+    },
+    RvError {
+        return_value: Type,
+        error:        Type,
+    },
+    ParamRvError {
+        parameter:    Type,
+        return_value: Type,
+        error:        Type,
+    },
+    ParamEvent {
+        parameter: Type,
+        event:     EventSchema,
+    },
+    RvEvent {
+        return_value: Type,
+        event:        EventSchema,
+    },
+    ParamRvEvent {
+        parameter:    Type,
+        return_value: Type,
+        event:        EventSchema,
+    },
+    ErrorEvent {
+        error: Type,
+        event: EventSchema,
+    },
+    ParamErrorEvent {
+        parameter: Type,
+        error:     Type,
+        event:     EventSchema,
+    },
```

Comment:
I think this is getting out of hand. I strongly recommend using instead

```
pub struct FunctionV3 {
   parameter: Option<Type>,
   return_value: Option<Type>,
   error: OPtion<Type>,
   event: Option<Type>
}
```
instead. And then have the property that at least one of them is not None (which we can check during deserialization).